### PR TITLE
added windows port

### PR DESCRIPTION
**Pull request by schoetbi**
Do you like to support windows?

### CODE REVIEW

**Comment by 7thSamurai** [comment]
Hey, I appreciate you taking the time to add in Windows support, but according to the official Windows API documentation, isn't `CryptGenRandom()` deprecated? Wouldn't it be better to use `BCryptGenRandom()` or something?

**Comment by schoetbi** [comment]
@7thSamurai I do not know if there are issues with the old API, it was an example I had, but if you want, I can modify the RNG to [BCryptGenRandom](https://learn.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgenrandom)

**Comment by 7thSamurai** [comment]
Yeah, I'd appreciate that. While it probably isn't a big deal to use the old method, I just figure that we might as well use the new version. 

Besides Microsoft does state that "Microsoft may remove this API in future releases.", so while that's unlikely I'd prefer not to risk any issues in the future. Thanks.

File: src/main.cpp [hunk 1]
```diff
@@ -74,24 +71,29 @@ int encode(Image &image, const std::array<std::uint8_t, 32> &password, const std
...
-    std::fill(padded_data + size, padded_data + size + left, left);
+    std::fill_n(padded_data.get() + size, left, left);
 
     // Pick a random offset inside the image to store the data
     std::uint32_t offset;
     Random random;
-    random.get(&offset, sizeof(offset));
+	if (!random.get(&offset, sizeof(offset)))
+	{
+		std::cout << "Unable to generate random number\n";
```
Comment (7thSamurai) [inline src/main.cpp]:
```suggestion
		std::cerr << "Unable to generate random number" << std::endl;
```
Please output errors to standard error, and use `std::endl` to make sure that the buffer is flushed.

File: src/main.cpp [hunk 2]
```diff
@@ -117,8 +119,11 @@ int encode(Image &image, const std::array<std::uint8_t, 32> &password, const std
 
     // Generate the Salt and IV
     std::uint8_t salt[16], iv[16];
-    random.get(salt, sizeof(salt));
-    random.get(iv,   sizeof(iv));
+	if (!random.get(salt, sizeof salt) || !random.get(iv, sizeof iv))
+	{
+		std::cout << "ERROR: Unable to generate random number\n";
```
Comment (7thSamurai) [inline src/main.cpp]:
```suggestion
		std::cerr << "ERROR: Unable to generate random number" << std::endl;
```
Please output errors to standard error, and use std::endl to make sure that the buffer is flushed.

File: src/main.cpp [hunk 3]
```diff
@@ -128,20 +133,20 @@ int encode(Image &image, const std::array<std::uint8_t, 32> &password, const std
 
     // Encrypt the header
     AES aes(key, iv);
-    auto encrypted_header = new std::uint8_t[sizeof(header)];
-    aes.cbc_encrypt(&header, sizeof(header), encrypted_header);
+    auto encrypted_header = std::make_unique<uint8_t>(sizeof Header);
```
Comment (7thSamurai) [inline src/main.cpp]:
```suggestion
    auto encrypted_header = std::make_unique<uint8_t[]>(sizeof Header);
```
Please use the array version of `std::make_unique`.

File: src/main.cpp [hunk 4]
```diff
@@ -128,20 +133,20 @@ int encode(Image &image, const std::array<std::uint8_t, 32> &password, const std
...
+    aes.cbc_encrypt(&header, sizeof(header), encrypted_header.get());
 
     // Encrypt the data
-    auto encrypted_data = new std::uint8_t[padded_size];
-    aes.cbc_encrypt(padded_data, padded_size, encrypted_data);
+    auto encrypted_data = std::make_unique<uint8_t>(padded_size);
```
Comment (7thSamurai) [inline src/main.cpp]:
```suggestion
    auto encrypted_data = std::make_unique<uint8_t[]>(padded_size);
```
Please use the array version of `std::make_unique`.

File: src/random.hpp [hunk 5]
```diff
@@ -2,15 +2,66 @@
...
+		file = fopen("/dev/urandom", "rb");
+	}
+	~Random() {
+		if (file)
+			fclose(file);
+	}
+
+	bool get(void* data, std::size_t size) {
+		fread(data, size, 1, file);
+		return true;
```
Comment (7thSamurai) [inline src/random.hpp]:
```suggestion
		return fread(data, size, 1, file) == size;
```
Check for error in `fread()`.

File: src/random.hpp [hunk 6]
```diff
@@ -2,15 +2,66 @@
...
+	bool get(void* data, std::size_t size) {
+		auto status = BCryptGenRandom(
+			NULL,
+			(BYTE*)data,
+			size,
+			BCRYPT_USE_SYSTEM_PREFERRED_RNG);
+
+		if (!BCRYPT_SUCCESS(status))
+		{
+			std::cerr << "Unable to generate random number\n";
```
Comment (7thSamurai) [inline src/random.hpp]:
```suggestion
			std::cerr << "Unable to generate random number" << std::endl;
```
Please use `std::endl` to make sure the buffer gets flushed.

File: src/random.hpp [hunk 7]
```diff
@@ -2,15 +2,66 @@
...
+			return false;
+		}
+		return true;		
+	}
 
 private:
-    FILE *file;
+	HCRYPTPROV hCryptProv = NULL;
+	LPCSTR UserName = "MyKeyContainer";
+
```
Comment (7thSamurai) [inline src/random.hpp]:
```suggestion
```
Remove old unneeded lines.

File: src/random.hpp [hunk 8]
```diff
@@ -2,15 +2,66 @@
...
+#elif defined(_WIN32)
 public:
-    Random();
-    ~Random();
+	Random() {
+
+	}
+	~Random() {
+		
+	}
```
Comment (7thSamurai) [inline src/random.hpp]:
```suggestion
	Random() = default;
	~Random() = default;
```
Use default constructors and destructors.

**Review by 7thSamurai — COMMENTED** [review]
Looking great! I left a couple of comments about a few issues that I want resolved before I merge this PR. Good work!

**Review by 7thSamurai — APPROVED** [review]
I've just reviewed your changes, and they all look great! I'll merge this PR in just a minute.

**Pull request merged**